### PR TITLE
metamorphic: fix tracking of external ingest keys with synthetic prefix

### LINKED ISSUE #3325: github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14090083?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/14090083?buildTab=artifacts#meta) on refs/heads/master @ [9e60abf53295](https://github.com/cockroachdb/pebble/commits/9e60abf53295):

```
db1.Set("qvkgrjxug", "cqaaqoezpsbccputhsf")
external0 = batch0.NewExternalObj()
db1.Get("hmkq@1")
db1.Get("qvkgrjxug")
db1.DeleteRange("hmkq@1", "qvkgrjxug")
db1.Delete("hmkq@1")
db1.SingleDelete("qvkgrjxug", false /* maybeReplaceDelete */)
db1.LogData("ntruc")
db1.Delete("hmkq@1")
db1.Set("idcbkgvfe@1", "fazcqqbonggict")
batch1 = db1.NewIndexedBatch()
db1.Merge("hmkq@1", "zwuluo")
batch1.Set("basnf@1", "xtu")
db1.Get("idcbkgvfe@1")
db1.Ingest(batch1)
db1.Get("jsbkdnb@1")
db1.Get("idcbkgvfe@1")
db1.Delete("basnf@1")
db1.Set("basnf@1", "ibxbcrzaovin")

--- FAIL: TestMeta (0.00s)
panic: invalid bounds ["jsbkdnb@1", "jsbkdnb@1") [recovered]
	panic: invalid bounds ["jsbkdnb@1", "jsbkdnb@1") [recovered]
	panic: invalid bounds ["jsbkdnb@1", "jsbkdnb@1")

goroutine 5 [running]:
testing.tRunner.func1.2({0xc92880, 0xc000052cc0})
	GOROOT/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x397
panic({0xc92880?, 0xc000052cc0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.generate.func1()
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:208 +0xa5
panic({0xc92880?, 0xc000052cc0?})
	GOROOT/src/runtime/panic.go:914 +0x21f
github.com/cockroachdb/pebble/metamorphic.(*generator).writerIngestExternalFiles(0xc0000f8000)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:1324 +0x1345
github.com/cockroachdb/pebble/metamorphic.generate(_, _, {{0x5, 0x5, 0x1, 0x0, 0x1, 0x2, 0x1, 0x2, ...}, ...}, ...)
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/generator.go:212 +0xd5e
github.com/cockroachdb/pebble/metamorphic.RunAndCompare(0xc0000ac820, {0x7ffeda25a93c, 0xf}, {0xc0000b8990, 0x3, 0x3})
	github.com/cockroachdb/pebble/metamorphic/external/com_github_cockroachdb_pebble/metamorphic/meta.go:199 +0x730
github.com/cockroachdb/pebble/internal/metamorphic.runTestMeta(0xc000072760?, 0x0)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:94 +0x239
github.com/cockroachdb/pebble/internal/metamorphic.TestMeta(0x0?)
	github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/meta_test.go:54 +0x15
testing.tRunner(0xc0000ac820, 0xdfce78)
	GOROOT/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
	GOROOT/src/testing/testing.go:1648 +0x3ad
```
<details><summary>Help</summary>
<p>
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by RaduBerinde**
Note : `require ExternalFile bounds without suffixes` commit is a separate PR #3357

#### metamorphic: track whether objects have range dels/keys

We use this information to select a Batch without range keys when
creating an external object (rather than ignoring them). We will use
the range dels to correctly generate synthetic suffix/prefix upfront
(instead of correcting them during `run()`, which makes tracking key
histories inaccurate).

We also clean up the object merging code a bit.

#### metamorphic: return []keyMeta instead of []*keyMeta

Change `SortedKeysForObj` and `inRangeKeys` to return a slice of
`keyMeta` instead of `*keyMeta`. This makes it more clear that we
don't modify them (and allows local modification as needed).

#### metamorphic: fix tracking of external ingest keys with synthetic prefix

This commit corrects the tracking of external ingestion keys in the
key manager when synthetic suffixes are used. This is important for
detecting SINGLEDEL conflicts correctly.

Fixes #3325

### CODE REVIEW

**Review by jbowens — APPROVED** [review]
:lgtm_strong: 

> Reviewed 5 of 5 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 6 of 6 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/3363#-:-Nrr3wDF6Fgn5eq3Uiho:b-uwkowd)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @itsbilal)

**Comment by RaduBerinde** [comment]
TFTR! Looking good so far:
```
1588 runs so far, 0 failures, over 3h49m20s
```

**Comment by RaduBerinde** [comment]
Will wait for a review on #3357 before merging.

**Pull request merged**
